core/src/main/java/com/homeaway/streamplatform/streamregistry/StreamRegistryApplication.java

```diff
@@ -174,18 +174,56 @@ public void run(final StreamRegistryConfiguration configuration, final Environme
-        StreamClientDao<Consumer> consumerDao = new ConsumerDaoImpl(managedProducer, managedKStreams, env, regionDao, infraManager, kafkaManager);
+        JerseyEnvironment jersey = environment.jersey();
+        jersey.register(new AbstractBinder() {
 
-        StreamResource streamResource = new StreamResource(streamDao, producerDao, consumerDao);
-        environment.jersey().register(streamResource);
-        environment.jersey().register(new RegionResource(regionDao));
+            @Override
+            protected void configure() {
+                bind(managedKStreams).to(ManagedKStreams.class);
+                bind(managedInfraManager).to(ManagedInfraManager.class);
+                bind(managedProducer).to(ManagedKafkaProducer.class);
+                bind(env).to(String.class).named("stream-registry-env");
+                bind(regionDao).to(RegionDao.class);
+                bind(streamValidator).to(StreamValidator.class);
+                bind(schemaManager).to(SchemaManager.class);
+
+                bind(configuration.getHealthCheckStreamConfig()).to(HealthCheckStreamConfig.class);
+                bind(metricRegistry).to(MetricRegistry.class);
+                bind(StreamRegistryHealthCheck.class).to(StreamRegistryHealthCheck.class);
+
+                bind(StreamDaoImpl.class).to(StreamDao.class).in(Singleton.class);
+                bind(ProducerDaoImpl.class).to(new TypeLiteral<StreamClientDao<Producer>>() {}).in(Singleton.class);
+                bind(ConsumerDaoImpl.class).to(new TypeLiteral<StreamClientDao<Consumer>>() {}).in(Singleton.class);
+                bind(StreamResource.class).to(StreamResource.class).in(Singleton.class);
+                bind(RegionResource.class).to(RegionResource.class).in(Singleton.class);
+
+            }
+        });
+
+        environment.jersey().register(StreamResource.class);
+        environment.jersey().register(RegionResource.class);
 
         environment.getApplicationContext().addServlet(PingServlet.class, "/ping");
-        StreamRegistryHealthCheck streamRegistryHealthCheck = new StreamRegistryHealthCheck(managedKStreams, streamResource, metricRegistry, configuration.getHealthCheckStreamConfig());
 
-        environment.healthChecks().register("streamRegistryHealthCheck", streamRegistryHealthCheck);
+        environment.jersey().register(new ApplicationEventListener() {
+
+            @Override
+            public void onEvent(ApplicationEvent event) {
+                if (event.getType() == ApplicationEvent.Type.INITIALIZATION_FINISHED) {
+                    ServiceLocator serviceLocator =
+                        ((ServletContainer) environment.getJerseyServletContainer()).getApplicationHandler().getServiceLocator();
+
+                    StreamRegistryHealthCheck streamRegistryHealthCheck = serviceLocator.getService(StreamRegistryHealthCheck.class);
+
+                    environment.healthChecks().register("streamRegistryHealthCheck", streamRegistryHealthCheck);
```

Comment:
AFAIK (might be wrong here) jersey expects its registration to happen during startup of the Application.  By deferring its registration in the event handler, are we creating possible timing issues ?

Suggestion: Move the registration inline. And make StreamRegistryHealthCheck stateful (e.g. it will be uninitialized).  In this event handler, change the state to be INITIALIZED.  If someone pings StreamRegistryHealthCheck while initializing, they will get a DOWN answer until it is INITIALIZED and passes its checks).

This may cause issues with the IT tests... but that's fine... we'll add something that waits appropriately in the IT setup() method to wait for healthcheck to be healthy.

Reply:
Came across [this](https://github.com/dropwizard/dropwizard/issues/862) moments back. Maybe will explore options sans HK2 ( inline with your feedback ).

Also will check how existing PR plays with `INITIALIZED` ( last I tested + debugged seems the service locator is instantiated post the servlet transition to `INITIALIZATION_FINISHED` state. i'll re-verify this and come back )

Reply:
nit - confirming that during `INITIALIZATION_START` & `INITIALIZATION_APP_FINISHED` - `((ServletContainer) environment.getJerseyServletContainer()).getApplicationHandler()` is `NULL` and thus we cannot get back `ServiceLocator`.

Reply:
[Here](https://gist.github.com/saketanisha/9406dd4b43b73e7ef00809931d84745d) is the trimmed down server startup logs.

For this experiment I removed the code inside `ApplicationEventListener` to register the health check - i.e. `environment.healthChecks().register("streamRegistryHealthCheck", streamRegistryHealthCheck);` and added a nit instrumentation to `INITIALIZATION_FINISHED` callback.

1. `/private` is registered [here](https://gist.github.com/saketanisha/9406dd4b43b73e7ef00809931d84745d#file-stream-registry-startup-log-L7).
2. `INITIALIZATION_FINISHED` callback is called [here](https://gist.github.com/saketanisha/9406dd4b43b73e7ef00809931d84745d#file-stream-registry-startup-log-L34).
3. `/private` servlet context is started [here](https://gist.github.com/saketanisha/9406dd4b43b73e7ef00809931d84745d#file-stream-registry-startup-log-L71) and before this it DW setup create `AdminEnvironment` for `/private` - see [here](https://gist.github.com/saketanisha/9406dd4b43b73e7ef00809931d84745d#file-stream-registry-startup-log-L62-L70) .
4. Finally the app http connector is bound to port 8080 and server is ready to server - [here](https://gist.github.com/saketanisha/9406dd4b43b73e7ef00809931d84745d#file-stream-registry-startup-log-L72-L73). 

I believe this change is inline with the servlet lifecycle. An observation is that its kind-of hooked with the `init` lifecycle method, i.e. semantic working of server is tightly couple with underlying servlet container.

As an experiment let me try out taking out the call back completely out of picture and drive this orchestration via single managed bean within DW that encapsulates the other building block of the system.

**_Note_**: please ignore [these](https://gist.github.com/saketanisha/9406dd4b43b73e7ef00809931d84745d#file-stream-registry-startup-log-L45-L54) error logs - the local kafka cluster not configured with infra bits.

Reply:
( forgot to mention above - during INITIALIZATION_START & `INITIALIZATION_APP_FINISHED` - `((ServletContainer) environment.getJerseyServletContainer()).getApplicationHandler()` is `NULL` )

Reply:
Removed dependency on HK2 and Jersey Container lifecycle.